src/cli/src/install/mod.rs

```diff
@@ -122,24 +116,27 @@ fn verify_checksum<B: AsRef<[u8]>>(buffer: B, checksum: &str) -> bool {
     &*buffer_checksum == checksum
 }
 
-pub fn install_bin<P: AsRef<Path>, B: AsRef<[u8]>>(
-    bin_dir: P,
-    name: &str,
-    bytes: B,
-) -> Result<(), CliError> {
+pub fn install_bin<P: AsRef<Path>, B: AsRef<[u8]>>(bin_path: P, bytes: B) -> Result<(), CliError> {
     use std::io::Write as _;
 
-    // Create bin_dir if it does not exist
-    let bin_dir = bin_dir.as_ref();
-    std::fs::create_dir_all(&bin_dir)?;
+    let bin_path = bin_path.as_ref();
 
-    // Install our package to `<bin_dir>/<name>`
-    let install_path = bin_dir.join(name);
-    let mut install_file = File::create(&install_path)?;
-    install_file.write_all(bytes.as_ref())?;
+    // Create directories to bin_path if they do not exist
+    if let Some(parent) = bin_path.parent() {
+        std::fs::create_dir_all(&parent)?;
+    }
+
+    // Write bin to temporary file
+    let tmp_dir = tempdir::TempDir::new("fluvio")?;
+    let tmp_path = tmp_dir.path().join("fluvio");
+    let mut tmp_file = File::create(&tmp_path)?;
+    tmp_file.write_all(bytes.as_ref())?;
```

Comment:
shouldn't use https://doc.rust-lang.org/std/fs/fn.copy.html ?

Reply:
This is not the same update strategy that we started out with. Originally, we were going to have the Fluvio CLI copy itself to another file and then invoke that file. Now, instead, we are simply writing the new downloaded file to a temporary directory and then renaming it to "overwrite" the original executable. This would be equivalent to `mv old new`. Importantly, it doesn't actually write over the old (currently executing) executable (it still has the same file descriptor, but now the other file has the file _name_).